File: day24.c
#include <assert.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct gate_t {
    char name[4];
    char op[4];
    char a[4], b[4];
    struct gate_t* inp[2];
    int out;
    int cache;
} gate_t;

gate_t gates[512];
int ngates;
gate_t* xyzgates[3][64];

bool panic = false;
int eval(gate_t* gate, int depth)
{
    depth++;
    if (depth > 100) {
        //puts("panic");
        panic = true;
        return 0;
    }
    //puts(gate->name);
    int value = gate->out ? gate->out : gate->cache;
    if (!value) {
        int a = eval(gate->inp[0], depth);
        int b = eval(gate->inp[1], depth);
        if (!strcmp(gate->op, "AND")) value = a & b;
        if (!strcmp(gate->op, "OR" )) value = a | b;
        if (!strcmp(gate->op, "XOR")) value = a ^ b;
        value++;
        gate->cache = value;
    }
    return value - 1;
}

int collect(int* results, char prefix)
{
    panic = false;
    for (int i = 0; i < ngates; i++) {
        gates[i].cache = 0;
    }
    int n;
    for (n = 0; n < 64; n++) {
        gate_t* gate = xyzgates[prefix - 'x'][n];
        if (!gate) break;
        results[n] = eval(gate, 0);
    }
    return n;
}

long long tonum(int* results, int n)
{
    long long num = 0;
    for (int i = 0; i < n; i++) {
        if (results[i]) num |= 1LL << i;
    }
    return num;
}

void tobin(int* bits, int n, long long value)
{
    for (int i = 0; i < n; i++) {
        bits[i] = (value >> i) & 1;
    }
}

void setnum(char prefix, int* b, int n)
{
    for (int i = 0; i < n; i++) {
        xyzgates[prefix - 'x'][i]->out = b[i] + 1;
    }
}

void swap(gate_t* a, gate_t* b)
{
    gate_t tmp = *a;
    *a = *b;
    *b = tmp;
}

int finderror(int nz, int nx, bool any)
{
    int lowest = 0;
    for (int trial = 0; trial < 100; trial++) {
        int x[64], y[64], z[64], w[64];
        for (int i = 0; i < nx; i++) {
            x[i] = rand() % 2;
            y[i] = rand() % 2;
        }
        setnum('x', x, nx);
        setnum('y', y, nx);
        collect(z, 'z');
        tobin(w, nz, tonum(x, nx) + tonum(y, nx));
        if (panic) {
            lowest = 1;
        } else {
            for (int i = 0; i < nz; i++) {
                if (z[i] != w[i]) {
                    if (!lowest || i + 1 < lowest) lowest = i + 1;
                    break;
                }
            }
        }
        if (any && lowest) break;
    }
    return lowest;
}

int cmpstr(const void* a, const void* b) { return strcmp(a, b); }

int main()
{
    FILE* file = fopen("day24.txt", "r");

    for (char line[32]; fgets(line, sizeof(line), file); ) {
        gate_t* gate = &gates[ngates];
        if (2 == sscanf(line, "%[a-z0-9]: %d", gate->name, &gate->out)) {
            strcpy(gate->op, "SET");
            gate->out++;
        } else if (4 == sscanf(line, "%s %s %s -> %s", gate->a, gate->op, gate->b, gate->name)) {
        } else continue;
        ngates++;
        if (gate->name[0] >= 'x' && gate->name[0] <= 'z') {
            xyzgates[gate->name[0] - 'x'][atoi(gate->name + 1)] = gate;
        }
    }

    for (int i = 0; i < ngates; i++) {
        if (!gates[i].a[0]) continue;
        for (int j = 0; j < ngates; j++) {
            if (!strcmp(gates[i].a, gates[j].name)) {
                gates[i].inp[0] = &gates[j];
            } else if (!strcmp(gates[i].b, gates[j].name)) {
                gates[i].inp[1] = &gates[j];
            }
        }
    }

    int x[64];
    //int y[64];
    int z[64];
    //int w[64];
    int nx = collect(x, 'x');
    //int ny = collect(y, 'y');
    int nz = collect(z, 'z');
    //tobin(w, nz, tonum(x, nx) + tonum(y, ny));

    //printf("%lld\n", tonum(x, nx));
    //printf("%lld\n", tonum(y, ny));
    printf("%lld\n", tonum(z, nz));
    //printf("%lld\n", tonum(w, nz));

    char solution[8][4];
    int nsolution = 0;

    while (true) {
        int bit = finderror(nz, nx, false);
        if (!bit) break;
        //printf("bit %d\n", bit - 1);

        gate_t* pair[2] = { 0 };
        for (int i = 0; i < ngates; i++) {
            gate_t* a = &gates[i];
            if (!a->inp[0]) continue;
            for (int j = i + 1; j < ngates; j++) {
                gate_t* b = &gates[j];
                if (!b->inp[0]) continue;
                swap(a, b);
                if (!finderror(bit, nx, true)) {
                    //printf("swap %s %s\n", a->name, b->name);
                    assert(!pair[0]);
                    pair[0] = a;
                    pair[1] = b;
                }
                swap(a, b);
            }
        }
        if (!pair[0]) {
            puts("failed");
            break;
        }
        swap(pair[0], pair[1]);
        strcpy(solution[nsolution++], pair[0]->name);
        strcpy(solution[nsolution++], pair[1]->name);
    }

    qsort(solution, nsolution, sizeof(solution[0]), cmpstr);
    for (int i = 0; i < nsolution; i++) {
        if (i > 0) printf(",");
        printf("%s", solution[i]);
    }
    printf("\n");
}

File: day24.lua
#!/usr/bin/env lua

local input = io.open("day24.txt"):read("a")

local gates = {}
local list = {}
local cache = {}

for name, value in input:gmatch("(%w+): (%d)") do
    local gate = { name = name, op = "SET", out = tonumber(value) }
    gates[name] = gate
end
for a, op, b, name in input:gmatch("(%w+) (%w+) (%w+) %-> (%w+)") do
    local gate = { name = name, op = op, inp = { a, b } }
    gates[name] = gate
    table.insert(list, name)
end
table.sort(list)

local panic = false
local function eval(gate, depth)
    depth = (depth or 0) + 1
    if depth > 100 then
        --print("panic")
        panic = true
        return 0
    end
    --print(gate.name)
    local value = gate.out or cache[gate]
    if not value then
        local a = eval(gates[gate.inp[1]], depth)
        local b = eval(gates[gate.inp[2]], depth)
        if gate.op == "AND" then value = a & b end
        if gate.op == "OR"  then value = a | b end
        if gate.op == "XOR" then value = a ~ b end
        cache[gate] = value
    end
    return value
end

local xyzname = { x = {}, y = {}, z = {} }
local function makename(prefix, index)
    local name = xyzname[prefix][index]
    if not name then
        name = string.format("%s%02d", prefix, index - 1)
        xyzname[prefix][index] = name
    end
    return name
end

local function collect(prefix)
    panic = false
    cache = {}
    local results = {}
    for i = 1, math.huge do
        local gate = gates[makename(prefix, i)]
        if not gate then break end
        results[i] = eval(gate)
    end
    return results
end

local function tonum(results)
    return tonumber(table.concat(results):reverse(), 2)
end

local function tobin(value)
    local bits = {}
    for i = 1, math.huge do
        bits[i] = value & 1
        value = value >> 1
        if value == 0 then break end
    end
    return bits
end

local function setnum(prefix, n, value)
    local b = tobin(value)
    for i = 1, n do
        gates[makename(prefix, i)].out = b[i] or 0
    end
end

local function swap(a, b)
    gates[a], gates[b] = gates[b], gates[a]
end

local function finderror(nz, nx, any)
    local lowest
    for trial = 1, 100 do
        local xnum = math.random((1 << nx) - 1)
        local ynum = math.random((1 << nx) - 1)
        setnum("x", nx, xnum)
        setnum("y", nx, ynum)
        local z = collect("z")
        local w = tobin(xnum + ynum)
        if panic then
            lowest = 0
        else
            for i = 1, nz do
                if z[i] ~= (w[i] or 0) then
                    lowest = math.min(i, lowest or math.huge)
                    break
                end
            end
        end
        if any and lowest then break end
    end
    return lowest
end

local x = collect("x")
--local y = collect("y")
local z = collect("z")
--local w = tobin(tonum(x) + tonum(y))

--print(tonum(x))
--print(tonum(y))
print(tonum(z))
--print(tonum(w))

local solution = {}

while true do
    local bit = finderror(#z, #x, false)
    if not bit then break end
    --print("bit", bit - 1)

    local pair = nil
    for i = 1, #list do
        for j = i + 1, #list do
            swap(list[i], list[j])
            if not finderror(bit, #x, true) then
                --print("swap", list[i], list[j])
                assert(not pair)
                pair = { list[i], list[j] }
            end
            swap(list[i], list[j])
        end
    end
    if not pair then
        print("failed")
        break
    end
    swap(pair[1], pair[2])
    table.insert(solution, pair[1])
    table.insert(solution, pair[2])
end

table.sort(solution)
print(table.concat(solution, ","))
